### PR TITLE
Feature/issue383 -- closes #383

### LINKED ISSUE #383: matrix transpose
Would be nice to have too.  R has [do_transpose](https://github.com/wch/r-source/blob/trunk/src/main/array.c#L1065-L1188).

### PR DESCRIPTION
**Pull request by eddelbuettel**
this should close #383:
- works on square and non-square integer and real matrices
- deals with `dimnames`
- will add one for strings as well, not sure I can be bothered for complex

### CODE REVIEW

**Comment by eddelbuettel** [comment]
Bump to @jjallaire @kevinushey -- please take a look when time permits.

This is an addition of three free functions (along with tests), and changes no interface.

**Comment by jjallaire** [comment]
LGTM however I don't know enough about the internals to fully provide the green light. @kevinushey ?

File: inst/include/Rcpp/vector/Matrix.h [hunk 1]
```diff
@@ -361,6 +360,102 @@ inline std::ostream &operator<<(std::ostream & s, const Matrix<RTYPE, StoragePol
...
+    R_xlen_t len = XLENGTH(x), l_1 = XLENGTH(x)-1;
+
+    // similar approach as in R: fill by in column, "accessing row-wise"
+    VECTOR s = VECTOR(r.get__());
+    for (R_xlen_t i = 0, j = 0; i < len; i++, j += nrow) {
+        if (j > l_1) j -= l_1;
+        s[i] = x[j];
+    }
+
+    // there must be a simpler, more C++-ish way for this ...
```
Comment (kevinushey) [inline inst/include/Rcpp/vector/Matrix.h]: You could do something like this:

```
SEXP oldDimNames = Rf_getAttrib(x, R_DimNamesSymbol);
if (!Rf_isNull(oldDimNames)) {
    Shield<SEXP> newDimNames(Rf_allocVector(VECSXP, 2));
    SET_VECTOR_ELT(newDimNames, 0, VECTOR_ELT(oldDimNames, 1));
    SET_VECTOR_ELT(newDimNames, 1, VECTOR_ELT(oldDimNames, 0));
    Rf_setAttrib(r, R_DimNamesSymbol, newDimNames);
}
```

Marginally cleaner?

File: inst/include/Rcpp/vector/Matrix.h [hunk 2]
```diff
@@ -361,6 +360,102 @@ inline std::ostream &operator<<(std::ostream & s, const Matrix<RTYPE, StoragePol
...
+    SEXP rnames = internal::DimNameProxy(x, 0);
+    SEXP cnames = internal::DimNameProxy(x, 1);
+    if (!Rf_isNull(rnames) || !Rf_isNull(cnames)) {
+        SEXP dimnames;
+        PROTECT(dimnames = Rf_allocVector(VECSXP, 2));
+        SET_VECTOR_ELT(dimnames, 0, cnames);
+        SET_VECTOR_ELT(dimnames, 1, rnames);
+        // do we need dimnamesnames ?
+        Rf_setAttrib(r, R_DimNamesSymbol, dimnames);
+        UNPROTECT(1); /* dimnames */
```
Comment (kevinushey) [inline inst/include/Rcpp/vector/Matrix.h]: Should use `Shield<SEXP>(Rf_allocVector(VECSXP, 2))` to ensure that destructors automatically take care of reference counting (don't need to worry about `UNPROTECT`)

File: inst/include/Rcpp/vector/Matrix.h [hunk 3]
```diff
@@ -361,6 +360,102 @@ inline std::ostream &operator<<(std::ostream & s, const Matrix<RTYPE, StoragePol
...
+    VECTOR s = VECTOR(r.get__());
+    for (R_xlen_t i = 0, j = 0; i < len; i++, j += nrow) {
+        if (j > l_1) j -= l_1;
+        s[i] = x[j];
+    }
+
+    // there must be a simpler, more C++-ish way for this ...
+    SEXP rnames = internal::DimNameProxy(x, 0);
+    SEXP cnames = internal::DimNameProxy(x, 1);
+    if (!Rf_isNull(rnames) || !Rf_isNull(cnames)) {
```
Comment (kevinushey) [inline inst/include/Rcpp/vector/Matrix.h]: Extract this to a helper function?

**Comment by kevinushey** [comment]
This is fine as-is but made some comments that might help clean up (code duplication, `Shield<SEXP>`)

File: inst/include/Rcpp/vector/Matrix.h (hunk 1 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: Nice. Anything that avoid `PROTECT` and `UNPROTECT` has my vote :)

File: inst/include/Rcpp/vector/Matrix.h (hunk 2 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: Yes, as just noted above -- very nice.

File: inst/include/Rcpp/vector/Matrix.h (hunk 3 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: Yes, could do.

**Comment by eddelbuettel** [comment]
Very helpful -- thanks for he comments, Kevin.  Will revise once more.

File: inst/include/Rcpp/vector/Matrix.h (hunk 3 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: On second look, maybe not so obvious.  The `typedef` are local; could do old-school via a macros but then with just three deployments... it seems fine _as is_.

File: inst/include/Rcpp/vector/Matrix.h (hunk 1 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: That's been committed.

File: inst/include/Rcpp/vector/Matrix.h (hunk 3 above)
Comment (eddelbuettel) [inline inst/include/Rcpp/vector/Matrix.h]: On third look, I think I have it worked out -- see ea52d607920499e938c9e33b4b94309d3b8a42fd and give it another look. 

Should be merge-ready now.

**Comment by eddelbuettel** [comment]
Bump.  @kevinushey if you can take another quick look...

**Comment by kevinushey** [comment]
Looks good to me!

**Pull request merged**